pinot-broker/src/main/java/org/apache/pinot/broker/routing/v2/RoutingManager.java

```diff
@@ -0,0 +1,481 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.routing.v2;
+
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.Nullable;
+import org.apache.helix.AccessOption;
+import org.apache.helix.BaseDataAccessor;
+import org.apache.helix.HelixConstants;
+import org.apache.helix.HelixDataAccessor;
+import org.apache.helix.HelixManager;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.model.ExternalView;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.apache.pinot.broker.broker.helix.ClusterChangeHandler;
+import org.apache.pinot.broker.routing.v2.instanceselector.InstanceSelector;
+import org.apache.pinot.broker.routing.v2.instanceselector.InstanceSelectorFactory;
+import org.apache.pinot.broker.routing.v2.segmentpruner.SegmentPruner;
+import org.apache.pinot.broker.routing.v2.segmentpruner.SegmentPrunerFactory;
+import org.apache.pinot.broker.routing.v2.segmentselector.SegmentSelector;
+import org.apache.pinot.broker.routing.v2.segmentselector.SegmentSelectorFactory;
+import org.apache.pinot.broker.routing.v2.timeboundary.TimeBoundaryInfo;
+import org.apache.pinot.broker.routing.v2.timeboundary.TimeBoundaryManager;
+import org.apache.pinot.common.config.TableConfig;
+import org.apache.pinot.common.config.TableNameBuilder;
+import org.apache.pinot.common.metadata.ZKMetadataProvider;
+import org.apache.pinot.common.metrics.BrokerMeter;
+import org.apache.pinot.common.metrics.BrokerMetrics;
+import org.apache.pinot.common.request.BrokerRequest;
+import org.apache.pinot.common.utils.CommonConstants;
+import org.apache.pinot.common.utils.CommonConstants.Helix.StateModel.RealtimeSegmentOnlineOfflineStateModel;
+import org.apache.pinot.common.utils.HashUtil;
+import org.apache.pinot.core.transport.ServerInstance;
+import org.apache.zookeeper.data.Stat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The {@code RoutingManager} manages the routing of all tables hosted by the broker instance.
+ */
+public class RoutingManager implements ClusterChangeHandler {
```

Comment:
With this new class, what happens to HelixExternalViewBasedRouiting? How is this different from HelixExternalViewBasedRouting? 

That class also implements ClusterChangeHandler interface. Is the plan to make this an abstract class and have HelixExternalViewBasedRouting extend it or do we plan to remove it?

Reply:
The implementation in this pr will replace everything under the package `org.apache.pinot.broker.routing` once we plug them in